File: android/app/src/main/java/bd/sicip/qavisit/domain/Scoring.kt
// visit scoring: category -> points, rank aggregation. category is picked manually at
// End tour (no more auto-derivation from dates). pure kotlin, no android deps.
package bd.sicip.qavisit.domain

// fixed points scale per category.
val POINTS: Map<String, Int> = mapOf(
    "A***" to 116,
    "A**+" to 112,
    "A**" to 100,
    "A++*" to 96,
    "A++" to 84,
    "A+*" to 80,
    "A+" to 68,
    "A*" to 64,
    "A" to 52,
    "B+" to 48,
    "B" to 36,
    "C+" to 32,
    "C" to 20,
    "D+" to 16,
    "D" to 4,
    "E" to 1,
    "N/A" to 0,
)

fun points(category: String): Int = POINTS[category] ?: 0

// category -> (days, nights) span it represents. single source for bill allowances: nights-away
// and food-days are both derived from this, not stored/edited separately -- see suggestedNights/
// suggestedFood below and CATEGORIES.md at repo root for the full table + formula.
val CATEGORY_SPANS: Map<String, Pair<Int, Int>> = mapOf(
    "A***" to (8 to 7),
    "A**+" to (7 to 7),
    "A**" to (7 to 6),
    "A++*" to (6 to 6),
    "A++" to (6 to 5),
    "A+*" to (5 to 5),
    "A+" to (5 to 4),
    "A*" to (4 to 4),
    "A" to (4 to 3),
    "B+" to (3 to 3),
    "B" to (3 to 2),
    "C+" to (2 to 2),
    "C" to (2 to 1),
    "D+" to (1 to 1),
    "D" to (1 to 0),
    "E" to (0 to 0),
    "N/A" to (0 to 0),
)

// accommodation nights = the category's night count, flat.
fun suggestedNights(category: String): Int = CATEGORY_SPANS[category]?.second ?: 0

// food-days = every night full + a half-day for each day beyond the last night (the day-1
// morning-out and the final travel-home day), same shape as BillMath's old span-default rule,
// now keyed off the category instead of raw dates.
fun suggestedFood(category: String): Double {
    val (d, n) = CATEGORY_SPANS[category] ?: (0 to 0)
    return n + 0.5 * (d - n)
}

// full explanation shown wherever the user picks/reads a category (dropdowns); stored value stays
// the bare code (POINTS key) -- this map is display-only.
val CATEGORY_LABELS: Map<String, String> = mapOf(
    "A***" to "A*** — 8D7N (116 pts)",
    "A**+" to "A**+ — 7D7N (112 pts)",
    "A**" to "A** — 7D6N (100 pts)",
    "A++*" to "A++* — 6D6N (96 pts)",
    "A++" to "A++ — 6D5N (84 pts)",
    "A+*" to "A+* — 5D5N (80 pts)",
    "A+" to "A+ — 5D4N (68 pts)",
    "A*" to "A* — 4D4N (64 pts)",
    "A" to "A — 4D3N (52 pts)",
    "B+" to "B+ — 3D3N (48 pts)",
    "B" to "B — 3D2N (36 pts)",
    "C+" to "C+ — 2D2N (32 pts)",
    "C" to "C — 2D1N (20 pts)",
    "D+" to "D+ — 1D1N (16 pts)",
    "D" to "D — 1 day / Dhaka non-metro (4 pts)",
    "E" to "E — Dhaka metro (1 pt)",
    "N/A" to "N/A — Additional (0 pts)",
)

// minimal shape scoring needs from a visit; keeps this file free of Room/android deps.
// done = visit.status == "done" -- only finished visits score, scheduled ones are worth 0.
data class VisitScore(val officerId: String, val category: String, val deleted: Boolean = false, val done: Boolean = true)

fun totalPoints(visits: List<VisitScore>): Int =
    visits.filter { !it.deleted && it.done }.sumOf { points(it.category) }

// officers ranked by summed points, lowest first (fewer points = better rank, see DECISIONS.md)
fun rank(visits: List<VisitScore>): List<Pair<String, Int>> =
    visits.filter { !it.deleted && it.done }
        .groupBy { it.officerId }
        .mapValues { (_, v) -> v.sumOf { points(it.category) } }
        .entries.sortedBy { it.value }
        .map { it.key to it.value }

// visits within a calendar month, for the home dashboard's "this month: n visits · m pts" line.
data class MonthVisit(val startDate: String, val category: String, val deleted: Boolean = false, val done: Boolean = true)

// yearMonth as "yyyy-MM" (java.time.YearMonth.toString() shape); startDate compared by its
// first 7 chars so this stays a plain string op, no date parsing needed.
fun monthSummary(visits: List<MonthVisit>, yearMonth: String): Pair<Int, Int> {
    val inMonth = visits.filter { !it.deleted && it.done }.filter { it.startDate.take(7) == yearMonth }
    return inMonth.size to inMonth.sumOf { points(it.category) }
}

File: web/src/lib/scoring.js
// visit scoring: category -> points, rank aggregation.
// ported 1:1 from android/app/src/main/java/bd/sicip/qavisit/domain/Scoring.kt

// fixed points scale per category. classic ladder (nights = days-1) vs plus ladder
// (nights >= days, i.e. an extra night beyond the classic span).
export const POINTS = {
  'A***': 116, 'A**+': 112, 'A**': 100, 'A++*': 96, 'A++': 84,
  'A+*': 80, 'A+': 68, 'A*': 64, 'A': 52, 'B+': 48, 'B': 36,
  'C+': 32, 'C': 20, 'D+': 16, 'D': 4, 'E': 1, 'N/A': 0,
}

export function points(category) {
  return POINTS[category] ?? 0
}

// category -> [days, nights] span it represents. single source for bill allowances.
export const CATEGORY_SPANS = {
  'A***': [8, 7], 'A**+': [7, 7], 'A**': [7, 6], 'A++*': [6, 6], 'A++': [6, 5],
  'A+*': [5, 5], 'A+': [5, 4], 'A*': [4, 4], 'A': [4, 3], 'B+': [3, 3], 'B': [3, 2],
  'C+': [2, 2], 'C': [2, 1], 'D+': [1, 1], 'D': [1, 0], 'E': [0, 0], 'N/A': [0, 0],
}

// accommodation nights = the category's night count, flat.
export function suggestedNights(category) {
  return CATEGORY_SPANS[category]?.[1] ?? 0
}

// food-days = every night full + a half-day for each day beyond the last night.
export function suggestedFood(category) {
  const [d, n] = CATEGORY_SPANS[category] ?? [0, 0]
  return n + 0.5 * (d - n)
}

// full explanation shown wherever the user picks/reads a category (dropdowns).
export const CATEGORY_LABELS = {
  'A***': 'A*** — 8D7N (116 pts)',
  'A**+': 'A**+ — 7D7N (112 pts)',
  'A**': 'A** — 7D6N (100 pts)',
  'A++*': 'A++* — 6D6N (96 pts)',
  'A++': 'A++ — 6D5N (84 pts)',
  'A+*': 'A+* — 5D5N (80 pts)',
  'A+': 'A+ — 5D4N (68 pts)',
  'A*': 'A* — 4D4N (64 pts)',
  'A': 'A — 4D3N (52 pts)',
  'B+': 'B+ — 3D3N (48 pts)',
  'B': 'B — 3D2N (36 pts)',
  'C+': 'C+ — 2D2N (32 pts)',
  'C': 'C — 2D1N (20 pts)',
  'D+': 'D+ — 1D1N (16 pts)',
  'D': 'D — 1 day / Dhaka non-metro (4 pts)',
  'E': 'E — Dhaka metro (1 pt)',
  'N/A': 'N/A — Additional (0 pts)',
}

// only done, non-deleted visits score (scheduled visits don't count yet). missing `done` = true.
export function totalPoints(visits) {
  return visits.filter((v) => !v.deleted && v.done !== false).reduce((sum, v) => sum + points(v.category), 0)
}

// officers ranked by summed points, fewest first (#1 = fewest points). returns [[officerId, points], ...]
export function rank(visits) {
  const byOfficer = new Map()
  for (const v of visits) {
    if (v.deleted || v.done === false) continue
    byOfficer.set(v.officerId, (byOfficer.get(v.officerId) ?? 0) + points(v.category))
  }
  return [...byOfficer.entries()].sort((a, b) => a[1] - b[1])
}

// visits within a calendar month (yearMonth = "yyyy-MM"), for "this month: n visits · m pts".
export function monthSummary(visits, yearMonth) {
  const inMonth = visits.filter((v) => !v.deleted && v.done !== false && v.startDate.slice(0, 7) === yearMonth)
  return [inMonth.length, inMonth.reduce((sum, v) => sum + points(v.category), 0)]
}

// "Last month" rank cutoff: last day of the month before `today` (iso "yyyy-MM-dd"). day 0 of
// the current month is the previous month's last day -- rolls Jan back into prior-year Dec 31,
// leap Feb included, for free (no manual month-length table). mirrors Rank.kt 1:1.
export function lastDayOfPreviousMonth(todayIso) {
  const [y, m] = todayIso.split('-').map(Number)
  return new Date(Date.UTC(y, m - 1, 0)).toISOString().slice(0, 10)
}
